### PR TITLE
Added "Gradient" trees/forests to compute derivatives

### PR DESCRIPTION
**Pull request by NiklasPfister**
Refactored the old LinearRegression trees into the tree type Gradient. This allows to directly compute gradients. I further also added a new criteria "Partial_quadratic" which allows to fit quadratics instead of just linear functions as with "Partial_linear" (the old "Linear_regression" criteria).

@svbrodersen It would be great if you could do a code review on this change.

### CODE REVIEW

File: src/adaXT/base_model.pyx [hunk 1]
```diff
@@ -62,19 +65,19 @@ class BaseModel:
                     self.leaf_builder_class = leaf_builder
                 else:
                     self.leaf_builder_class = LeafBuilderRegression
-            elif tree_type == "LinearRegression":
+            elif tree_type == "Gradient":
```
Comment (svbrodersen) [inline src/adaXT/base_model.pyx:68]: Would we not want a check both for the linear and the new quadratic?

File: src/adaXT/base_model.pyx [hunk 2]
```diff
@@ -62,19 +65,19 @@ class BaseModel:
...
                 if predict:
                     self.predict_class = predict
                 else:
-                    self.predict_class = PredictLinearRegression
+                    self.predict_class = PredictLocalPolynomial
                 if criteria:
                     self.criteria_class = criteria
                 else:
-                    self.criteria_class = Linear_regression
+                    self.criteria_class = Partial_quadratic
```
Comment (svbrodersen) [inline src/adaXT/base_model.pyx:76]: The quadratic fits quadratically, no?

Would it not be better suited for the "Gradient" default to make use of the Partial_linear, and then having some other naming for the quadratic?

File: src/adaXT/base_model.pyx [hunk 3]
```diff
@@ -62,19 +65,19 @@ class BaseModel:
...
                 if leaf_builder:
                     self.leaf_builder_class = leaf_builder
                 else:
-                    self.leaf_builder_class = LeafBuilderLinearRegression
+                    self.leaf_builder_class = LeafBuilderPartialQuadratic
```
Comment (svbrodersen) [inline src/adaXT/base_model.pyx:80]: Same about naming convention.

File: src/adaXT/criteria/criteria.pxd [hunk 4]
```diff
@@ -32,7 +32,7 @@ cdef class Criteria:
         Parameters
         ----------
         indices: int[:]
-        the indices of a given node
+       the indices of a given node
```
Comment (svbrodersen) [inline src/adaXT/criteria/criteria.pxd:35]: No need to change the indentation here. Now it is off.

File: src/adaXT/criteria/criteria.pyx [hunk 5]
```diff
@@ -467,10 +467,10 @@ cdef class Squared_error(Criteria):
         square_err = cur_sum/obs_weight - mu*mu
         return square_err
 
-cdef class Linear_regression(Criteria):
+cdef class Partial_linear(Criteria):
 
     # Custom mean function, such that we don't have to loop through twice.
-    cdef (double, double) custom_mean(self, int[:] indices):
+    cdef (double, double) _custom_mean(self, int[:] indices):
```
Comment (svbrodersen) [inline src/adaXT/criteria/criteria.pyx:473]: If we see them as only private functions I think the general convention is to use a two underscores (__) for private methods.

File: src/adaXT/criteria/criteria.pyx [hunk 6]
```diff
@@ -483,10 +483,11 @@ cdef class Linear_regression(Criteria):
 
         return ((sumX / (<double> length)), (sumY/ (<double> length)))
 
-    cdef (double, double) theta(self, int[:] indices):
+    cdef (double, double) _theta(self, int[:] indices):
```
Comment (svbrodersen) [inline src/adaXT/criteria/criteria.pyx:486]: Private method.

File: src/adaXT/criteria/criteria.pyx [hunk 7]
```diff
@@ -541,9 +542,107 @@ cdef class Linear_regression(Criteria):
             int i, length
 
         length = indices.shape[0]
-        theta0, theta1 = self.theta(indices)
+        theta0, theta1 = self._theta(indices)
         cur_sum = 0.0
         for i in range(length):
             step_calc = self.y[indices[i]] - theta0 - theta1 * self.x[indices[i], 0]
-            cur_sum += step_calc*step_calc
+            cur_sum += step_calc ** 2
```
Comment (svbrodersen) [inline src/adaXT/criteria/criteria.pyx:549]: Generally it is less efficient to do something ** 2 instead of doing something*something. The same convention can be seen in scikit learns: https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/tree/_criterion.pyx.

This answer goes more in depth: https://stackoverflow.com/a/72153092.

File: src/adaXT/criteria/criteria.pyx [hunk 8]
```diff
@@ -541,9 +542,107 @@ cdef class Linear_regression(Criteria):
...
+        cdef:
+            double sumXsq, sumX, sumY
+            int i
+            int length = indices.shape[0]
+        sumX = 0.0
+        sumXsq = 0.0
+        sumY = 0.0
+        for i in range(length):
+            sumX += self.x[indices[i], 0]
+            sumXsq += self.x[indices[i], 0] ** 2
```
Comment (svbrodersen) [inline src/adaXT/criteria/criteria.pyx:565]: Same with the square here.

File: src/adaXT/criteria/criteria.pyx [hunk 9]
```diff
@@ -541,9 +542,107 @@ cdef class Linear_regression(Criteria):
...
-        theta0, theta1 = self.theta(indices)
+        theta0, theta1 = self._theta(indices)
         cur_sum = 0.0
         for i in range(length):
             step_calc = self.y[indices[i]] - theta0 - theta1 * self.x[indices[i], 0]
-            cur_sum += step_calc*step_calc
+            cur_sum += step_calc ** 2
+        return cur_sum
+
+cdef class Partial_quadratic(Criteria):
```
Comment (svbrodersen) [inline src/adaXT/criteria/criteria.pyx:552]: Might be an idea to create a "Partial" parent class, which inherits from the Criteria class. That way the _custom_mean does not have to be created twice.

File: src/adaXT/criteria/criteria.pyx [hunk 10]
```diff
@@ -541,9 +542,107 @@ cdef class Linear_regression(Criteria):
...
+        for i in range(length):
+            X_diff = self.x[indices[i], 0] - muX
+            Xsq_diff = self.x[indices[i], 0] ** 2 - muXsq
+            Y_diff = self.y[indices[i]] - muY
+            covXXsq += X_diff * Xsq_diff
+            varX += X_diff ** 2
+            varXsq += Xsq_diff ** 2
+            covXY += X_diff * Y_diff
+            covXsqY += Xsq_diff * Y_diff
+        det = varX * varXsq - covXXsq ** 2
```
Comment (svbrodersen) [inline src/adaXT/criteria/criteria.pyx:610]: Same with the square in the section above here.

File: src/adaXT/criteria/criteria.pyx [hunk 11]
```diff
@@ -541,9 +542,107 @@ cdef class Linear_regression(Criteria):
...
+            int i, length
+
+        length = indices.shape[0]
+        theta0, theta1, theta2 = self._theta(indices)
+        cur_sum = 0.0
+        for i in range(length):
+            step_calc = self.y[indices[i]] - theta0
+            step_calc -= theta1 * self.x[indices[i], 0]
+            step_calc -= theta2 * self.x[indices[i], 0] ** 2
+            cur_sum += step_calc ** 2
```
Comment (svbrodersen) [inline src/adaXT/criteria/criteria.pyx:647]: And square above here.

File: src/adaXT/leaf_builder/leaf_builder.pxd [hunk 12]
```diff
@@ -19,7 +19,7 @@ cdef class LeafBuilderClassification(LeafBuilder):
         double[::1] classes
         int n_classes
 
-    cdef double[::1] __get_mean(self, int[::1] indices)
+    cdef double[::1] _get_mean(self, int[::1] indices)
```
Comment (svbrodersen) [inline src/adaXT/leaf_builder/leaf_builder.pxd:22]: why is this changed to a single underscore?

File: src/adaXT/leaf_builder/leaf_builder.pxd [hunk 13]
```diff
@@ -32,7 +32,7 @@ cdef class LeafBuilderClassification(LeafBuilder):
 
 cdef class LeafBuilderRegression(LeafBuilder):
 
-    cdef double __get_mean(self, int[::1] indices)
+    cdef double _get_mean(self, int[::1] indices)
```
Comment (svbrodersen) [inline src/adaXT/leaf_builder/leaf_builder.pxd:35]: and here?

File: src/adaXT/leaf_builder/leaf_builder.pyx [hunk 14]
```diff
@@ -139,11 +141,99 @@ cdef class LeafBuilderLinearRegression(LeafBuilderRegression):
...
 
-        return LinearRegressionLeafNode(leaf_id, indices, depth, impurity,
-                                        weighted_samples, mean, parent, theta0,
-                                        theta1)
+        return LocalPolynomialLeafNode(leaf_id, indices, depth, impurity,
+                                       weighted_samples, mean, parent, theta0,
+                                       theta1, theta2)
+
+
+cdef class LeafBuilderPartialQuadratic(LeafBuilderRegression):
```
Comment (svbrodersen) [inline src/adaXT/leaf_builder/leaf_builder.pyx:155]: Same with creating a parent class might be an idea.

File: src/adaXT/predict/predict.pyi [hunk 15]
```diff
@@ -20,8 +20,16 @@ class PredictRegression(Predict):
 
     pass
 
-class PredictLinearRegression(Predict):
+class PredictLocalPolynomial(Predict):
```
Comment (svbrodersen) [inline src/adaXT/predict/predict.pyi:23]: From what I can tell this is not the default in the base_model.pyx

File: src/adaXT/base_model.pyx [hunk 16]
```diff
@@ -62,19 +65,19 @@ class BaseModel:
...
 
         else:
```
Comment (svbrodersen) [inline src/adaXT/base_model.pyx:82]: Quantile is missing a check setting its defaults.

File: src/adaXT/predict/predict.pyx [hunk 17]
```diff
@@ -214,8 +221,16 @@ cdef class PredictLinearRegression(PredictRegression):
                     cur_node = cur_node.left_child
                 else:
                     cur_node = cur_node.right_child
-            Y[i] = cur_node.theta0 + cur_node.theta1*X[i, 0]
-        return Y
+            ind = 0
+            for oo in order:
```
Comment (svbrodersen) [inline src/adaXT/predict/predict.pyx:225]: loop variable oo might as well be defined as an integer under the cdef.

File: src/adaXT/predict/predict.pyx [hunk 18]
```diff
@@ -214,8 +221,16 @@ cdef class PredictLinearRegression(PredictRegression):
...
+                if oo == 0:
+                    deriv_mat[i, ind] = cur_node.theta0 + cur_node.theta1*X[i, 0] + cur_node.theta2*X[i, 0] ** 2
+                if oo == 1:
```
Comment (svbrodersen) [inline src/adaXT/predict/predict.pyx:228]: should be changed to elif statements, cant both be 0 and 1.

File: src/adaXT/predict/predict.pyx [hunk 19]
```diff
@@ -214,8 +221,16 @@ cdef class PredictLinearRegression(PredictRegression):
...
+                    deriv_mat[i, ind] = cur_node.theta1 + 2.0 * cur_node.theta2*X[i, 0]
+                if oo == 2:
```
Comment (svbrodersen) [inline src/adaXT/predict/predict.pyx:230]: same here

File: src/adaXT/predict/predict.pyx [hunk 20]
```diff
@@ -191,19 +189,28 @@ cdef class PredictRegression(Predict):
...
+            int i, cur_split_idx, n_obs, ind
             double cur_threshold
             object cur_node
-            double[:] Y
+            double[:, ::1] deriv_mat
+
+        if "order" not in kwargs.keys():
+            order = [0, 1, 2]
+        else:
+            order = np.array(kwargs['order'], ndmin=1, dtype='int')
```
Comment (svbrodersen) [inline src/adaXT/predict/predict.pyx:207]: might be an idea to add a check that size < 3. Currently we allow for duplicates. Not necessary.

**Review by svbrodersen — COMMENTED** [review]
Overall seems to look fine. Some questions about defaults and other things I think could be changed.

File: src/adaXT/base_model.pyx (hunk 1 above)
Comment (NiklasPfister) [reply src/adaXT/base_model.pyx:68]: I think the Partial_linear criteria is too niche to deserve its own default

File: src/adaXT/base_model.pyx (hunk 2 above)
Comment (NiklasPfister) [reply src/adaXT/base_model.pyx:76]: see above

File: src/adaXT/criteria/criteria.pyx (hunk 5 above)
Comment (NiklasPfister) [reply src/adaXT/criteria/criteria.pyx:451]: changed everything to double dash now

File: src/adaXT/criteria/criteria.pyx (hunk 7 above)
Comment (NiklasPfister) [reply src/adaXT/criteria/criteria.pyx:527]: Good point, thanks. Adjusted now.

File: src/adaXT/criteria/criteria.pyx (hunk 9 above)
Comment (NiklasPfister) [reply src/adaXT/criteria/criteria.pyx:530]: Might be a good idea if we extend further, currently the custom mean is different for both so lets just keep it as is for now.

File: src/adaXT/predict/predict.pyi (hunk 15 above)
Comment (NiklasPfister) [reply src/adaXT/predict/predict.pyi:23]: Double checked but seems to be correct

File: src/adaXT/base_model.pyx (hunk 16 above)
Comment (NiklasPfister) [reply src/adaXT/base_model.pyx:82]: Should also be set correctly?

File: src/adaXT/predict/predict.pyx (hunk 18 above)
Comment (NiklasPfister) [reply src/adaXT/predict/predict.pyx:228]: done

File: src/adaXT/predict/predict.pyx (hunk 20 above)
Comment (NiklasPfister) [reply src/adaXT/predict/predict.pyx:207]: done
